In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("day_wise_aqi_data.csv")

In [3]:
df

,date,state,area,number_of_monitoring_stations,prominent_pollutants,aqi_value,air_quality_status,unit
0,19-06-2025,Uttar Pradesh,Agra,5.0,"O3,PM2.5,PM10",49,Good,number_of_monitoring_stations in Absolute Numb...
1,19-06-2025,Karnataka,Bagalkot,1.0,PM10,46,Good,number_of_monitoring_stations in Absolute Numb...
2,19-06-2025,Maharashtra,Akola,1.0,PM10,26,Good,number_of_monitoring_stations in Absolute Numb...
3,19-06-2025,Rajasthan,Alwar,1.0,CO,76,Satisfactory,number_of_monitoring_stations in Absolute Numb...
4,19-06-2025,Andhra Pradesh,Amaravati,1.0,PM10,66,Satisfactory,number_of_monitoring_stations in Absolute Numb...
...,...,...,...,...,...,...,...,...
425274,01-05-2015,Uttar Pradesh,Kanpur,NaN,PM10,175,Moderate,number_of_monitoring_stations in Absolute Numb...
425275,01-05-2015,Uttar Pradesh,Agra,NaN,PM10,179,Moderate,number_of_monitoring_stations in Absolute Numb...
425276,01-05-2015,Tamil Nadu,Chennai,NaN,CO,87,Satisfactory,number_of_monitoring_stations in Absolute Numb...
425277,01-05-2015,Gujarat,Ahmedabad,NaN,PM2.5,168,Moderate,number_of_monitoring_stations in Absolute Numb...


In [4]:
print("Number of Columns(fields):",len(df.columns))
print("Name Of Columns:",df.columns)

Number of Columns(fields): 8
Name Of Columns: Index(['date', 'state', 'area', 'number_of_monitoring_stations',
       'prominent_pollutants', 'aqi_value', 'air_quality_status', 'unit'],
      dtype='object')


In [5]:
print("data types of columns:\n",df.dtypes)

data types of columns:
 date                              object
state                             object
area                              object
number_of_monitoring_stations    float64
prominent_pollutants              object
aqi_value                          int64
air_quality_status                object
unit                              object
dtype: object


#### as clearly shown from above description that we have to modify 2 fields i,e date and number_of_monitoring_stations .

In [5]:
df["date"] = pd.to_datetime(df["date"],format = "%d-%m-%Y")

In [6]:
df

,date,state,area,number_of_monitoring_stations,prominent_pollutants,aqi_value,air_quality_status,unit
0,2025-06-19,Uttar Pradesh,Agra,5.0,"O3,PM2.5,PM10",49,Good,number_of_monitoring_stations in Absolute Numb...
1,2025-06-19,Karnataka,Bagalkot,1.0,PM10,46,Good,number_of_monitoring_stations in Absolute Numb...
2,2025-06-19,Maharashtra,Akola,1.0,PM10,26,Good,number_of_monitoring_stations in Absolute Numb...
3,2025-06-19,Rajasthan,Alwar,1.0,CO,76,Satisfactory,number_of_monitoring_stations in Absolute Numb...
4,2025-06-19,Andhra Pradesh,Amaravati,1.0,PM10,66,Satisfactory,number_of_monitoring_stations in Absolute Numb...
...,...,...,...,...,...,...,...,...
425274,2015-05-01,Uttar Pradesh,Kanpur,NaN,PM10,175,Moderate,number_of_monitoring_stations in Absolute Numb...
425275,2015-05-01,Uttar Pradesh,Agra,NaN,PM10,179,Moderate,number_of_monitoring_stations in Absolute Numb...
425276,2015-05-01,Tamil Nadu,Chennai,NaN,CO,87,Satisfactory,number_of_monitoring_stations in Absolute Numb...
425277,2015-05-01,Gujarat,Ahmedabad,NaN,PM2.5,168,Moderate,number_of_monitoring_stations in Absolute Numb...


In [7]:
df.dtypes

date                             datetime64[ns]
state                                    object
area                                     object
number_of_monitoring_stations           float64
prominent_pollutants                     object
aqi_value                                 int64
air_quality_status                       object
unit                                     object
dtype: object

In [9]:
#------------------>

In [8]:
df["Year"] = df["date"].dt.year
df["month"] = df["date"].dt.month
df["day"] = df["date"].dt.day

In [11]:
df

,date,state,area,number_of_monitoring_stations,prominent_pollutants,aqi_value,air_quality_status,unit,Year,month,day
0,2025-06-19,Uttar Pradesh,Agra,5.0,"O3,PM2.5,PM10",49,Good,number_of_monitoring_stations in Absolute Numb...,2025,6,19
1,2025-06-19,Karnataka,Bagalkot,1.0,PM10,46,Good,number_of_monitoring_stations in Absolute Numb...,2025,6,19
2,2025-06-19,Maharashtra,Akola,1.0,PM10,26,Good,number_of_monitoring_stations in Absolute Numb...,2025,6,19
3,2025-06-19,Rajasthan,Alwar,1.0,CO,76,Satisfactory,number_of_monitoring_stations in Absolute Numb...,2025,6,19
4,2025-06-19,Andhra Pradesh,Amaravati,1.0,PM10,66,Satisfactory,number_of_monitoring_stations in Absolute Numb...,2025,6,19
...,...,...,...,...,...,...,...,...,...,...,...
425274,2015-05-01,Uttar Pradesh,Kanpur,NaN,PM10,175,Moderate,number_of_monitoring_stations in Absolute Numb...,2015,5,1
425275,2015-05-01,Uttar Pradesh,Agra,NaN,PM10,179,Moderate,number_of_monitoring_stations in Absolute Numb...,2015,5,1
425276,2015-05-01,Tamil Nadu,Chennai,NaN,CO,87,Satisfactory,number_of_monitoring_stations in Absolute Numb...,2015,5,1
425277,2015-05-01,Gujarat,Ahmedabad,NaN,PM2.5,168,Moderate,number_of_monitoring_stations in Absolute Numb...,2015,5,1


#### Null  Handling

In [9]:
df.isnull().sum()

date                               0
state                              0
area                               0
number_of_monitoring_stations    141
prominent_pollutants               0
aqi_value                          0
air_quality_status                 0
unit                               0
Year                               0
month                              0
day                                0
dtype: int64

In number_of_monitoring_stations there are 141 null values so we can assume that
there must be 1 monitoring_stations should be there to find out calculation
so we replace null by 1.


In [10]:
df["number_of_monitoring_stations"] = df["number_of_monitoring_stations"].fillna(1)

In [ ]:
df["number_of_monitoring_stations"]=df["number_of_monitoring_stations"].replace({0:1})

In [ ]:
df.isnull().sum()

In [11]:
def set_season(month) :
    if month in [12 , 1, 2] :
        return "Winter"
    elif month in [3,4,5]:
        return "Summer"
    elif month in [6,7,8,9] :
        return "Monsoon" 
    else :
        return "Post Monsoon"

df["season"] = df["month"].apply(set_season)

In [ ]:
df

In [ ]:
def validate_aqi_category(aqi):
    if aqi <= 50 :
        return "Good"
    elif aqi <= 100 :
        return "Satisfactory"
    elif aqi <= 200 :
        return "Moderate"
    elif aqi <= 300 :
        return "Poor"
    elif aqi <= 400 :
        return "Very Poor"
    else :
        return "Severe"

df["aqi_parameter"] = df["aqi_value"].apply(validate_aqi_category)
    

In [13]:
df["aqi_status_check"] = df["air_quality_status"] == df["aqi_parameter"]

In [14]:
df["aqi_status_check"].value_counts()

aqi_status_check
True     425011
False       268
Name: count, dtype: int64

In [ ]:
AS clearly shown From above figures that the original data set holds some wrong values in column
"air_quality_status" and we created one new column in which all value are right ,so we can delete
"air_quality_status".

In [15]:
df = df.drop(columns = "air_quality_status")

In [16]:
df = df.drop(columns = ["unit","aqi_status_check"])

In [ ]:
df

In [17]:
df["prominent_pollutants"] = df["prominent_pollutants"].str.split(",")

In [ ]:
df

In [18]:
df = df.explode("prominent_pollutants")

In [19]:
df["prominent_pollutants"] = df["prominent_pollutants"].str.strip()

In [ ]:
df

In [20]:
def determine_category(pollutant):
    if pollutant in ["PM2.5","PM10"]:
                 return 'Particulate'
    elif pollutant in ['CO', 'SO2', 'NO2','O3']:
                 return 'Gaseous'
    else:
                 return 'Secondary' 


df["pollutant Category"] = df["prominent_pollutants"].apply(determine_category)

In [ ]:
df

In [21]:
df.groupby("season")["aqi_value"].mean()

season
Monsoon          74.282410
Post Monsoon    146.117285
Summer          115.286574
Winter          155.031974
Name: aqi_value, dtype: float64

In [ ]:
df.groupby("season")["aqi_value"].agg(["mean","max","sum"])

#### Now we create a table (data frame)  in which we use group-by to group season and state

In [22]:
state_season_df = df.groupby(["season","state"]).agg( aqi_value = ("aqi_value","mean")).reset_index()

In [23]:
year_pollutant_df = df.groupby(["Year","prominent_pollutants"]).agg( aqi_value = ("aqi_value","mean")).reset_index()

In [24]:
year_pollutant_df.duplicated().sum()

np.int64(0)

In [25]:
state_season_df.duplicated().sum()

np.int64(0)

In [34]:
pip install pymysql sqlalchemy

Note: you may need to restart the kernel to use updated packages.


In [61]:
state_season_df

,season,state,aqi_value
0,Monsoon,Andaman and Nicobar Islands,69.909091
1,Monsoon,Andhra Pradesh,58.674800
2,Monsoon,Arunachal Pradesh,32.351351
3,Monsoon,Assam,55.896768
4,Monsoon,Bihar,86.345516
...,...,...,...
122,Winter,Telangana,115.793814
123,Winter,Tripura,203.268293
124,Winter,Uttar Pradesh,204.722743
125,Winter,Uttarakhand,111.652705


In [26]:
df.to_csv("AQI_Cleaned_Data.csv",index = False)

In [27]:
state_season_df.to_csv("State_By_Season_Cleaned_Data.csv",index = False)

In [28]:
year_pollutant_df.to_csv("Pollutant_Over_Years_Cleaned_Data.csv",index = False)

In [38]:
from sqlalchemy import create_engine
from urllib.parse import quote_plus


username = "root"
host = "localhost"
port = "3306"
password = quote_plus("bhanu@123")
database = "AQI_DataBase"

engine  = create_engine(f"mysql+pymysql://{username}:{password}@{host}:{port}/{database}")

table_name = "AQI_data"  
df.to_sql(table_name, engine, if_exists="replace", index=False)




C:\Users\Cybrom\AppData\Local\Temp\ipykernel_1556\117348402.py:14: UserWarning: The provided table name 'AQI_data' is not found exactly as such in the database after writing the table, possibly due to case sensitivity issues. Consider using lower case table names.
  df.to_sql(table_name, engine, if_exists="replace", index=False)


483820

'aaaaaaaaaaaaaaaaaaaaaaaaaaa9999 uuuuuuuuuuuuuuuuuuuuuuuuuu'

In [ ]:
def pollutant_type(p):
    if p in ['PM2.5', 'PM10']:
        return 'Particulate'
    elif p in ['CO', 'SO2', 'NO2']:
        return 'Gaseous'
    else:
        return 'Secondary'

df['pollutant_type'] = df['prominent_pollutants'].apply(pollutant_type)

In [ ]:
state_season_aqi = df.groupby(['state' , 'season']).agg(a = ("aqi_value","mean")).reset_index()

In [ ]:
state_season_aqi

In [ ]:
pip install pymysql sqlalchemy

In [ ]:
df.isnull().sum()

In [ ]:
df["number_of_monitoring_stations"] = df["number_of_monitoring_stations"].fillna(1)

In [ ]:
df["number_of_monitoring_stations"] = pd.to_numeric(
                                  df["number_of_monitoring_stations"], errors = "coerce")
df["aqi_value"] = pd.to_numeric(
                                  df["aqi_value"], errors = "coerce")

In [ ]:
df = df.dropna(subset = ["number_of_monitoring_stations",'aqi_value'])

In [ ]:
def set_season(x):
    if x in [12,1,2]:
        return "winter"
    elif x in [3,4,5] :
        return "summer"
    elif x in [6,7,8,9] :
        return "monsoon"
    else :
        return "post-monsoon"

df["Season"] = df["month"].apply(set_season)

In [ ]:
colmns = ["state","area"]
df[colmns] = df[colmns].apply(lambda x : x.str.strip().str.title() )
df[colmns]

In [ ]:
df

In [ ]:
df.groupby(['state','Season']).agg(aqi_vvvv=("aqi_value" ,"mean") ,aqivvvv=("aqi_value" ,"max"))

In [ ]:
df.groupby(["state","Season"]).agg(avg_aqi = ("aqi_value","mean"))

In [ ]:
df

In [ ]:
def validate_aqi_status(aqi):
    if aqi <= 50:
        return 'Good'
    elif aqi <= 100:
        return 'Satisfactory'
    elif aqi <= 200:
        return 'Moderate'
    elif aqi <= 300:
        return 'Poor'
    elif aqi <= 400:
        return 'Very Poor'
    else:
        return 'Severe'

df["Validate_status"] = df["aqi_value"].apply(validate_aqi_status)
    

In [ ]:
df["Check Status"] = df["air_quality_status"] == df["Validate_status"]

In [ ]:
df["Check Status"].value_counts()

In [ ]:
df[df["Check Status"] == False]

In [ ]:
df['prominent_pollutants'] = df['prominent_pollutants'].str.split(',')

In [ ]:
df

In [ ]:
df['prominent_pollutants']

In [ ]:
df = df.explode("prominent_pollutants")

In [ ]:
df

In [ ]:
df['prominent_pollutants'] = df['prominent_pollutants'].str.strip()

In [66]:
df

,date,state,area,number_of_monitoring_stations,prominent_pollutants,aqi_value,Year,month,day,season,aqi_parameter,pollutant Category
0,2025-06-19,Uttar Pradesh,Agra,5.0,O3,49,2025,6,19,Monsoon,Good,Gaseous
0,2025-06-19,Uttar Pradesh,Agra,5.0,PM2.5,49,2025,6,19,Monsoon,Good,Particulate
0,2025-06-19,Uttar Pradesh,Agra,5.0,PM10,49,2025,6,19,Monsoon,Good,Particulate
1,2025-06-19,Karnataka,Bagalkot,1.0,PM10,46,2025,6,19,Monsoon,Good,Particulate
2,2025-06-19,Maharashtra,Akola,1.0,PM10,26,2025,6,19,Monsoon,Good,Particulate
...,...,...,...,...,...,...,...,...,...,...,...,...
425274,2015-05-01,Uttar Pradesh,Kanpur,1.0,PM10,175,2015,5,1,Summer,Moderate,Particulate
425275,2015-05-01,Uttar Pradesh,Agra,1.0,PM10,179,2015,5,1,Summer,Moderate,Particulate
425276,2015-05-01,Tamil Nadu,Chennai,1.0,CO,87,2015,5,1,Summer,Satisfactory,Gaseous
425277,2015-05-01,Gujarat,Ahmedabad,1.0,PM2.5,168,2015,5,1,Summer,Moderate,Particulate


In [57]:
from sqlalchemy import create_engine
from urllib.parse import quote_plus

username = "root"
password = quote_plus("bhanu@123")
host = "localhost"
database = "airquality"
port = "3306"

engine = create_engine(f"mysql+pymysql://{username}:{password}@{host}:{port}/{database}")


In [58]:
table_name = "facttable"   # choose any table name
df.to_sql(table_name, engine, if_exists="replace", index=False)

483820

In [60]:
pd.read_sql("SELECT * FROM facttable LIMIT 5", engine)

,date,state,area,number_of_monitoring_stations,prominent_pollutants,aqi_value,Year,month,day,season,aqi_parameter,pollutant Category
0,2025-06-19,Uttar Pradesh,Agra,5.0,O3,49,2025,6,19,Monsoon,Good,Gaseous
1,2025-06-19,Uttar Pradesh,Agra,5.0,PM2.5,49,2025,6,19,Monsoon,Good,Particulate
2,2025-06-19,Uttar Pradesh,Agra,5.0,PM10,49,2025,6,19,Monsoon,Good,Particulate
3,2025-06-19,Karnataka,Bagalkot,1.0,PM10,46,2025,6,19,Monsoon,Good,Particulate
4,2025-06-19,Maharashtra,Akola,1.0,PM10,26,2025,6,19,Monsoon,Good,Particulate


In [62]:
table_name = "year_pollutant"   # choose any table name
df.to_sql(table_name, engine, if_exists="replace", index=False)

483820

In [63]:
table_name = "year_pollutant"   # choose any table name
year_pollutant_df.to_sql(table_name, engine, if_exists="replace", index=False)

80

In [64]:
table_name = "state_season"   # choose any table name
state_season_df.to_sql(table_name, engine, if_exists="replace", index=False)

127

In [65]:
pd.read_sql("Show tables", engine)

,Tables_in_airquality
0,dim_location
1,dim_pollutant
2,facttable
3,facttable_new
4,state_season
5,year_pollutant
